### DISCUSSION TITLE
Direct buffer support with standard parallel algorithms

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1493] [Q&A] [answered]
Hello, 

thank you again for amazing work.

I wanted to ask about new capabilities of offloading standard C++ parallel algorithms. I appreciate the innovative approach of integrating these algorithms into a SYCL environment, which significantly simplifies the development process for parallel applications.

However, how I can to use these algorithms in usual SYCL programs? I use SYCL buffers, and I manage memory using buffer and accessor paradigms typical in SYCL applications. While I understand that AdaptiveCpp transforms memory allocations to SYCL Unified Shared Memory (USM) for compatibility with standard algorithms, I am interested in a more direct integration with SYCL buffers without the intermediary step of converting to USM pointers.

Are there current or planned features in AdaptiveCpp that allow for using C++ standard parallel algorithms directly with SYCL buffers?
If not, could you suggest the most efficient method to use these algorithms with SYCL buffers, perhaps via buffer to USM conversions?

**Comment by illuhad** [comment, marked as answer]
> Are there current or planned features in AdaptiveCpp that allow for using C++ standard parallel algorithms directly with SYCL buffers?

No. Supporting SYCL buffers would have meant effectively implementing these algorithms twice. SYCL buffers are rarely useful compared to USM, and we see most new production code in SYCL using USM instead of buffers. Buffers can add substantial latency overhead compared to USM.

> If not, could you suggest the most efficient method to use these algorithms with SYCL buffers, perhaps via buffer to USM conversions?

AdaptiveCpp supports buffer-USM interop. It guarantees that all buffers will always use USM pointers under the hood, so you can always extract them. However, using this feature correctly requires some knowledge of AdaptiveCpp SYCL runtime internals, because if you make modifications to buffer data without the buffer being aware of the changes, the buffer's implicit data transfer logic will not do the right thing anymore.
This blog post explains some aspects of how to do it: https://adaptivecpp.github.io/hipsycl/extension/hipsycl-091-buffer-usm-interop/

**Reply by blinkfrog** [reply]
Thank you for your reply. This is understandable.

Factually, I just need a parallel primitives library, like that SyclParallelSTL, which would work with a modern SYCL 2020 implementations. I suppose, SyclParallelSTL won't work as is with AdaptiveCpp and oneAPI/DPC++ (especially on Windows, but I even haven't tried to build it), this is why I am trying to find any workarounds.

What is worse, my code should be fully portable and work with both AdaptiveCpp and oneAPI/DPC++, but I could use a conditional compiling and use oneDPL in oneAPI and something else (I hoped for this STL support) in AdaptiveCpp, although this isn't elegant.
